retrieval/discovery/file.go

```diff
@@ -198,11 +197,10 @@ func fileSource(filename string, i int) string {
 	return fmt.Sprintf("%s:%d", filename, i)
 }
 
-// Stop implements the TargetProvider interface.
-func (fd *FileDiscovery) Stop() {
+// stop shuts down the file watcher.
+func (fd *FileDiscovery) stop() {
 	log.Debugf("Stopping file discovery for %s...", fd.paths)
 
-	fd.done <- struct{}{}
 	// Closing the watcher will deadlock unless all events and errors are drained.
```

Comment:
Ugh, unrelated, but didn't notice this before - does this mean that there's a race condition between the drain here and the closing below (if new events happen after draining, but just before closing, it deadlocks)? Doesn't seem like there's a way to fix it though?